File: python/Linked List/add_two_numbers.py
from typing import Optional

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        """
        Add two numbers represented by linked lists (digits in reverse order).
        Approach:
        - Traverse both lists, summing corresponding digits and carry.
        - Create new nodes for each digit of the result.
        - Continue until both lists and carry are exhausted.
        Time Complexity: O(max(m, n)), where m and n are the lengths of the lists
        Space Complexity: O(max(m, n)) for the result list
        """
        dummy = ListNode()
        cur = dummy
        carry = 0

        while l1 or l2 or carry:
            v1 = l1.val if l1 else 0
            v2 = l2.val if l2 else 0

            #new digit
            val = v1 + v2 + carry
            
            carry = val // 10
            val = val % 10
            cur.next = ListNode(val)

            #update pointers
            cur = cur.next
            l1 = l1.next if l1 else None
            l2 = l2.next if l2 else None

        return dummy.next

# -------------------
# Helper Functions for Testing
# -------------------
def list_to_linkedlist(lst):
    dummy = ListNode()
    cur = dummy
    for val in lst:
        cur.next = ListNode(val)
        cur = cur.next
    return dummy.next

def linkedlist_to_list(node):
    result = []
    while node:
        result.append(node.val)
        node = node.next
    return result

# -------------------
# Test Cases
# -------------------
def test_add_two_numbers():
    s = Solution()
    # Test case 1: Basic
    l1 = list_to_linkedlist([2,4,3])
    l2 = list_to_linkedlist([5,6,4])
    result = s.addTwoNumbers(l1, l2)
    assert linkedlist_to_list(result) == [7,0,8]
    # Test case 2: Different lengths
    l1 = list_to_linkedlist([9,9,9,9,9,9,9])
    l2 = list_to_linkedlist([9,9,9,9])
    result = s.addTwoNumbers(l1, l2)
    assert linkedlist_to_list(result) == [8,9,9,9,0,0,0,1]
    # Test case 3: One list is empty
    l1 = list_to_linkedlist([])
    l2 = list_to_linkedlist([1,2,3])
    result = s.addTwoNumbers(l1, l2)
    assert linkedlist_to_list(result) == [1,2,3]
    # Test case 4: Both lists empty
    l1 = list_to_linkedlist([])
    l2 = list_to_linkedlist([])
    result = s.addTwoNumbers(l1, l2)
    assert linkedlist_to_list(result) == []
    # Test case 5: Carry at the end
    l1 = list_to_linkedlist([5])
    l2 = list_to_linkedlist([5])
    result = s.addTwoNumbers(l1, l2)
    assert linkedlist_to_list(result) == [0,1]
    print("All test cases passed!")

if __name__ == "__main__":
    test_add_two_numbers()

File: C++/Linked List/add_two_numbers.cpp
#include <bits/stdc++.h>  // includes all standard C++ libraries
using namespace std;

/**
 * Definition for singly-linked list.
 */
struct ListNode {
    int val;             // value of this node
    ListNode *next;      // pointer to the next node

    // Default constructor: sets val=0 and next=nullptr
    ListNode() : val(0), next(nullptr) {}

    // Constructor with just a value
    ListNode(int x) : val(x), next(nullptr) {}

    // Constructor with a value and pointer to next node
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
public:
    /**
     * Adds two numbers represented by two linked lists.
     * Digits are stored in reverse order.
     *
     * Time Complexity: O(max(m, n))
     * - because we traverse both lists once, where m and n are their lengths
     *
     * Space Complexity: O(max(m, n))
     * - because the result list can be as long as the longer input list (+1 if carry remains)
     */
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        // Create a dummy node to simplify result list construction
        ListNode* dummy = new ListNode();
        ListNode* cur = dummy;   // pointer to the current position in the result list
        int carry = 0;          // variable to hold carry from previous sum

        // Loop runs as long as either list has nodes left or there is a carry
        while (l1 != nullptr || l2 != nullptr || carry != 0) {
            int v1 = (l1 != nullptr) ? l1->val : 0;  // get value of current l1 node or 0 if l1 is exhausted
            int v2 = (l2 != nullptr) ? l2->val : 0;  // get value of current l2 node or 0 if l2 is exhausted

            int val = v1 + v2 + carry;    // sum current digits and carry
            carry = val / 10;            // compute new carry
            val = val % 10;             // keep only the single digit

            // Create a new node with the computed digit and link it to the result
            cur->next = new ListNode(val);

            cur = cur->next;  // move to the newly added node

            // Move to next nodes in l1 and l2 if possible
            if (l1 != nullptr) l1 = l1->next;
            if (l2 != nullptr) l2 = l2->next;
        }

        // Return the result list, skipping the dummy head
        return dummy->next;
    }
};
